### PR TITLE
New Regulation Endpoints, updated tests

### PR DESCRIPTION
**Pull request by juettemann**

### CODE REVIEW

File: lib/EnsEMBL/REST/Controller/Regulatory.pm [hunk 1]
```diff
@@ -23,6 +23,7 @@ use Moose;
 use namespace::autoclean;
 use Try::Tiny;
 use Bio::EnsEMBL::Utils::Scalar qw/check_ref/;
+use Data::Dumper;
```
Comment (magaliruffier) [inline lib/EnsEMBL/REST/Controller/Regulatory.pm]: not needed any more

File: lib/EnsEMBL/REST/Controller/Regulatory.pm [hunk 2]
```diff
@@ -36,29 +37,131 @@ application/json
...
+    try {
+      $microarrays = $c->model('Regulatory')->list_all_microarrays();
+    }catch {
+      $c->go('ReturnError', 'from_ensembl', [qq{$_}]) if $_ =~ /STACK/;
+      $c->go('ReturnError', 'custom', [qq{$_}]);
+    };
+    $self->status_ok($c, entity => $microarrays);
+}
+# stash array name
+sub microarray :Chained('species') PathPart('microarray') CaptureArgs(1) ActionClass('REST') {
```
Comment (nerdstrike) [inline lib/EnsEMBL/REST/Controller/Regulatory.pm]: Strictly speaking you could drop the ActionClass and fold microarray_get() into this method. There's no endpoint to it, so it doesn't need to act like one. Pending it working of course.

File: lib/EnsEMBL/REST/Controller/Regulatory.pm (hunk 1 above)
Comment (nerdstrike) [inline lib/EnsEMBL/REST/Controller/Regulatory.pm]: Redundant in production code.

File: lib/EnsEMBL/REST/Controller/Regulatory.pm [hunk 3]
```diff
@@ -36,29 +37,131 @@ application/json
...
+#protect against missing variables
 
-sub species: Chained('/') PathPart('regulatory') CaptureArgs(1) {
+sub species: Chained('/') PathPart('regulatory/species') CaptureArgs(1) {
   my ( $self, $c, $species) = @_;
   $c->stash(species => $species);
+
+  unless (defined $species) { $c->go('ReturnError', 'custom', [qq{Species must be provided as part of the URL.}])} 
+  $c->stash(species => $species);
+#  $c->log->debug(Dumper "S: $species");
```
Comment (magaliruffier) [inline lib/EnsEMBL/REST/Controller/Regulatory.pm]: not needed anymore

**Comment by lairdm** [comment]
Can you take the Data::Dumper and commented out debugging lines out please.

File: lib/EnsEMBL/REST/Controller/Regulatory.pm [hunk 4]
```diff
@@ -36,29 +37,131 @@ application/json
...
+# /regulatory/species/:species/id/:id
+sub id: Chained('species') PathPart('id') Args(1) ActionClass('REST') {
+  my ( $self, $c, $id) = @_;
 
-sub id: Chained('species') PathPart('') Args(1) ActionClass('REST') {
-  my $self = shift;
-  my $c    = shift;
-  my $id   = shift;
+  if(! defined $id){
+    $c->go('ReturnError', 'custom', [qq{Ensembl Stable ID  must be provided as part of the URL.}]);
```
Comment (magaliruffier) [inline lib/EnsEMBL/REST/Controller/Regulatory.pm]: not sure if the extra spacing is needed here

File: lib/EnsEMBL/REST/Controller/Regulatory.pm [hunk 5]
```diff
@@ -36,29 +37,131 @@ application/json
...
+  } 
+  if($id !~ /ENSR\d{11}/){
+    $c->go('ReturnError', 'custom', [qq{Ensembl Regulation Stable IDs  have the format ENSR12345678901.}])
```
Comment (magaliruffier) [inline lib/EnsEMBL/REST/Controller/Regulatory.pm]: or here

File: lib/EnsEMBL/REST/Controller/Regulatory.pm [hunk 6]
```diff
@@ -36,29 +37,131 @@ application/json
...
+  }
+}
+
+sub id_GET {
+  my ($self, $c, $id) = @_;
   my $regf;
-#  my $pops = $c->request->param('pops');
 
+#  $c->log->debug(Dumper "ID: $id");
```
Comment (magaliruffier) [inline lib/EnsEMBL/REST/Controller/Regulatory.pm]: remove debug

File: lib/EnsEMBL/REST/Model/Regulatory.pm [hunk 7]
```diff
@@ -35,28 +35,174 @@ sub build_per_context_instance {
...
-    Catalyst::Exception->throw("$regf_id not found for $species");
+    Catalyst::Exception->throw("$rf_id not found for $species");
+  }
+  my $result = $rf->summary_as_hash();
+
+  my $activity = $c->request->param('activity');
+  if($activity == 1) {
+    my $data; 
+    for my $ra(@{$rf->regulatory_activity}) {
+      $data->{$ra->epigenome->production_name} = $ra->activity;
```
Comment (nerdstrike) [inline lib/EnsEMBL/REST/Model/Regulatory.pm]: Courageous chaining unless you're sure that activity, epigenome and production_name will always be set

File: lib/EnsEMBL/REST/Controller/Regulatory.pm [hunk 8]
```diff
@@ -36,29 +37,131 @@ application/json
...
+  my ($self, $c, $probe_name) = @_;
+  if(! defined $probe_name){
+    $c->go('ReturnError', 'custom', [qq{Probe name must be provided as part of the URL.}]);
+  } 
+}
+sub microarray_probe_GET {
+  my ($self, $c, $probe_name) = @_;
+
+  my $probe_info;
+  $c->log->debug(Dumper "P: $probe_name");
```
Comment (magaliruffier) [inline lib/EnsEMBL/REST/Controller/Regulatory.pm]: remove debug

File: t/regulatory.t [hunk 9]
```diff
@@ -1,5 +1,5 @@
 # Copyright [1999-2015] Wellcome Trust Sanger Institute and the EMBL-European Bioinformatics Institute 
-# Copyright [2016-2017] EMBL-European Bioinformatics Institute
```
Comment (nerdstrike) [inline t/regulatory.t]: Put that back to 2017

File: lib/EnsEMBL/REST/Model/Regulatory.pm [hunk 10]
```diff
@@ -35,28 +35,174 @@ sub build_per_context_instance {
...
+  my $reg_build   = $reg_build_a->fetch_current_regulatory_build;
+  my $epigenomes  = $reg_build->get_all_Epigenomes;
+
+  my $result = [];
+  for my $eg (@$epigenomes) {
+    my $data = {};
+#    $data->{efo_id}             = $eg->efo_id;
+    $data->{gender}             = $eg->gender;
+    $data->{name}               = $eg->production_name;
+    #$data->{production_name}    = $eg->production_name;
```
Comment (magaliruffier) [inline lib/EnsEMBL/REST/Model/Regulatory.pm]: do the commented out lines need to stay?

File: lib/EnsEMBL/REST/Model/Regulatory.pm [hunk 11]
```diff
@@ -35,28 +35,174 @@ sub build_per_context_instance {
...
+          my $transcript = $tr_a->fetch_by_stable_id($tm->display_id);
+         $hash->{gene}->{stable_id}     = $transcript->get_Gene()->stable_id();
+         $hash->{gene}->{external_name} = $transcript->get_Gene()->external_name();
+        }
+        push(@$transcripts, $hash)
+      }
+      $features->{transcripts} = $transcripts;
+    }
+
+#    my $flag_all_matches = $c->request->param('all_matches');
```
Comment (magaliruffier) [inline lib/EnsEMBL/REST/Model/Regulatory.pm]: should probably go if it is not implemented

File: t/regulatory.t [hunk 12]
```diff
@@ -1,5 +1,5 @@
...
+# Copyright [2016] EMBL-European Bioinformatics Institute
```
Comment (magaliruffier) [inline t/regulatory.t]: you've lost the copyright here

File: t/regulatory.t [hunk 13]
```diff
@@ -33,21 +33,61 @@ use Bio::EnsEMBL::Test::TestUtils;
...
+
+my $microarray_vendor = '/regulatory/species/homo_sapiens/microarray/HumanWG_6_V2/vendor/illumina/?content-type=application/json';
+$output = {
+  format => 'EXPRESSION',
+  name   => 'HumanWG_6_V2',
+  class  => 'ILLUMINA_WG',
+  type   => 'OLIGO'
+};
+$json = json_GET($regulatory_feature,'GET Information about a specific microarray');
+#eq_or_diff($json, $output, 'GET Information about a specific microarray');
```
Comment (magaliruffier) [inline t/regulatory.t]: you're just ignoring the test here

File: t/regulatory.t [hunk 14]
```diff
@@ -33,21 +33,61 @@ use Bio::EnsEMBL::Test::TestUtils;
...
+$output = [ {
+    seq_region_name => '2',
+    Location        => 'chromosome:GRCh38:2:1:242193529:1',
+    Description     => 'null',
+    Name            => 'HumanWG_6_V2',
+    end             => 68735253,
+    start           => 68735204
+  } ];
+$json = json_GET($regulatory_feature,'GET Information about a specific probe');
+#eq_or_diff($json, $output, 'GET Information about a specific probe');
```
Comment (magaliruffier) [inline t/regulatory.t]: and here

**Comment by magaliruffier** [comment]
also looks like the overlap endpoint is broken for probe features, not sure if it is related to your changes or the code revert from regulation?

File: lib/EnsEMBL/REST/Model/Regulatory.pm [hunk 15]
```diff
@@ -35,28 +35,156 @@ sub build_per_context_instance {
...
+  }
+  my $result = $rf->summary_as_hash();
+
+  my $activity = $c->request->param('activity');
+  if($activity == 1) {
+    my $data; 
+    for my $ra(@{$rf->regulatory_activity}) {
+      $data->{$ra->epigenome->production_name} = $ra->activity;
+    }
+    $result ->{activity} = $data;
```
Comment (lairdm) [inline lib/EnsEMBL/REST/Model/Regulatory.pm]: needless space before the ->

File: lib/EnsEMBL/REST/Model/Regulatory.pm [hunk 16]
```diff
@@ -35,28 +35,156 @@ sub build_per_context_instance {
...
+  my $probe = $probe_adaptor->fetch_by_array_probe_probeset_name( $microarray, $probe_name );
+  my $probe_features = $probe_feature_adaptor->fetch_all_by_Probe($probe);
+
+  my $result = [];
+  for my $pf (@$probe_features) {
+    my $features = {};
+    $features->{name}            = $microarray ;
+    $features->{description}     = $pf->probe->description;
+    # Attaches whole chromosome instead of feature
+    # $features->{Location}        = $pf->seqname;
```
Comment (lairdm) [inline lib/EnsEMBL/REST/Model/Regulatory.pm]: Commented out for a reason? If not needed, remove please.

File: lib/EnsEMBL/REST/Model/Regulatory.pm [hunk 17]
```diff
@@ -35,28 +35,156 @@ sub build_per_context_instance {
...
+
+
+    my $flag_transcript = $c->request->param('transcript');
+    my $flag_gene       = $c->request->param('gene');
+    
+    # Linked transcripts
+    if($flag_transcript == 1){
+      my $transript_mappings = $pf->get_all_Transcript_DBEntries();
+      my $transcripts = [];
+      for my $tm (@$transript_mappings) {
```
Comment (lairdm) [inline lib/EnsEMBL/REST/Model/Regulatory.pm]: Should this not be foreach?

Comment (nerdstrike) [inline lib/EnsEMBL/REST/Model/Regulatory.pm]: for and foreach are syntactic sugar. They're both valid

**Comment by magaliruffier** [comment]
Travis failures:

t/regulatory.t .............. 1/? 
#   Failed test 'GET list of all epigenomes'
#   at t/regulatory.t line 49.
# +----+-----+----+----------------------------------+
# | Elt|Got  | Elt|Expected                          |
# +----+-----+----+----------------------------------+
# *   0|[]   *   0|[                                 *
# |    |     *   1|  {                               *
# |    |     *   2|    efo_id => 'efo:EFO_0002067',  *
# |    |     *   3|    gender => 'female',           *
# |    |     *   4|    name => 'K562',               *
# |    |     *   5|    scientific_name => 'K562'     *
# |    |     *   6|  }                               *
# |    |     *   7|]                                 *
# +----+-----+----+----------------------------------+
Use of uninitialized value $activity in numeric eq (==) at /home/travis/build/Ensembl/ensembl-rest/lib/EnsEMBL/REST/Model/Regulatory.pm line 52.
Use of uninitialized value $activity in numeric eq (==) at /home/travis/build/Ensembl/ensembl-rest/lib/EnsEMBL/REST/Model/Regulatory.pm line 52.
Use of uninitialized value $activity in numeric eq (==) at /home/travis/build/Ensembl/ensembl-rest/lib/EnsEMBL/REST/Model/Regulatory.pm line 52.
# Looks like you failed 1 test of 6.


and

You have not specified and Array objects, this will result in multiple/redundant queries based on the array_chip_id
You should pass Array objects to speed up this process at /home/travis/build/Ensembl/ensembl-rest/ensembl-funcgen/modules/Bio/EnsEMBL/Funcgen/Probe.pm line 170.

#   Failed test 'Getting probe_feature as JSON'
#   at /home/travis/build/Ensembl/ensembl-rest/t/lib/RestHelper.pm line 49.

#   Failed test 'Getting probe_feature as JSON'
#   at /home/travis/build/Ensembl/ensembl-rest/t/lib/RestHelper.pm line 49.

File: lib/EnsEMBL/REST/Model/Regulatory.pm (hunk 7 above)
Comment (juettemann) [inline lib/EnsEMBL/REST/Model/Regulatory.pm]: Braveheart shall be my name! On a more serious note: Yes, those are required values in the DB.

**Comment by magaliruffier** [comment]
info.t is now failing, the number of analyses expected is incorrect because you updated the test data

**Comment by magaliruffier** [comment]
Travis log is available here: https://s3.amazonaws.com/archive.travis-ci.org/jobs/206561110/log.txt
the test suite is passing now

**Pull request merged**